libmultilabel/metrics.py

```diff
@@ -31,68 +26,75 @@ class MultiLabelMetrics():
-        y_pred = np.vstack(self.y_pred)
-        report_dict = classification_report(y_true, y_pred > self.metric_threshold, output_dict=True, zero_division=0)
-        result = {
-            'Micro-Precision': report_dict['micro avg']['precision'],
-            'Micro-Recall': report_dict['micro avg']['recall'],
-            'Micro-F1': report_dict['micro avg']['f1-score'],
-            'Macro-F1': report_dict['macro avg']['f1-score'],
-            'Another-Macro-F1': another_macro_f1(
-                macro_prec=report_dict['macro avg']['precision'],
-                macro_rec=report_dict['macro avg']['recall'])  # caml's macro-f1
-        }
-        # add metrics like P@k, R@k to the result dict
+        y_pred_pos = y_pred > self.metric_threshold
+
+        n_eval = len(y_true)
+        self.n_eval += n_eval
+        self.multilabel_confusion_matrix += multilabel_confusion_matrix(y_true, y_pred_pos)
+
         scores = precision_recall_at_ks(y_true, y_pred, top_ks=self.top_ks)
-        precision_recall_monitored = [
-            metric for metric in self.monitor_metrics if metric in scores]
-        result.update({metric: scores[metric] for metric in precision_recall_monitored})
-        self.cached_results = result
-
-    def get_y_pred(self):
-        """Convert 3D array (shape: number of batches * batch_size * number of classes
-        to 2D array (shape: number of samples * number of classes).
-        """
-        return np.vstack(self.y_pred)
+        for metric in self.prec_recall_metrics:
+            self.metric_stats[metric] += (scores[metric] * n_eval)
 
-    def get_metric_dict(self, use_cache=False):
-        """Evaluate or get score dictionary from cache.
+    def get_metric_dict(self):
+        """Get evaluation results."""
 
-        Args:
-            threshold (float, optional): threshold to evaluate precision, recall, and f1 score. Defaults to 0.5.
-            use_cache (bool, optional): return cached results or not. Defaults to False.
-        """
-        if not use_cache:
-            self.eval()
-        return self.cached_results
+        result = {}
+        cm = self.multilabel_confusion_matrix
+        cum_cm = cm.sum(axis=0)
+        cum_tp, cum_fp, cum_fn = cum_cm[1,1], cum_cm[0,1], cum_cm[1,0]
+        micro_precision = cum_tp / (cum_tp + cum_fp + 1e-10)
+        micro_recall = cum_tp / (cum_tp + cum_fn + 1e-10)
+
+        ml_tp, ml_fp, ml_fn = cm[:,1,1], cm[:,0,1], cm[:,1,0]
+        labelwise_precision = ml_tp / (ml_tp + ml_fp + 1e-10)
+        labelwise_recall = ml_tp / (ml_tp + ml_fn + 1e-10)
+        macro_precision = labelwise_precision.mean()
+        macro_recall = labelwise_recall.mean()
+
+        result = {
+            'Micro-Precision': micro_precision,
+            'Micro-Recall': micro_recall,
+            'Micro-F1': f1(micro_precision, micro_recall),
+            'Macro-F1': f1(labelwise_precision, labelwise_recall).mean(),
+            'Another-Macro-F1': f1(macro_precision, macro_recall),
```

Comment:
Should keep the earlier comment to explain why we need another Macro. Indeed can expand it to cite the "Macro-F1 and macro-f1" paper. See caml.py now